File: dart/test/evaluator_test.dart
import 'dart:io';

import 'package:collection/collection.dart';
import 'package:eyamluate/eval/eval.dart';
import 'package:eyamluate/eval/evaluator.dart';
import 'package:eyamluate/yaml/decoder.dart';
import 'package:eyamluate/yaml/yaml.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

class _Testcase {
  String? yamlInput;
  Value? wantValue;
  bool? wantError;
}

void main() {
  final testcases = <String, _Testcase>{};
  final testdataDir =
      Directory(p.join(Directory.current.path, "test", "testdata"));
  testdataDir.listSync(recursive: true, followLinks: true).forEach((e) {
    if (e is! File) return;
    final path = p.canonicalize(e.path);
    if (e.path.endsWith(".in.yaml")) {
      final key = path.substring(0, path.length - ".in.yaml".length);
      if (!testcases.containsKey(key)) {
        testcases[key] = _Testcase();
      }
      testcases[key]!.yamlInput = e.readAsStringSync();
      return;
    }
    if (e.path.endsWith(".want.yaml")) {
      final key = path.substring(0, path.length - ".want.yaml".length);
      if (!testcases.containsKey(key)) {
        testcases[key] = _Testcase();
      }
      final d = Decoder().decode(DecodeInput(yaml: e.readAsStringSync()));
      if (d.isError) {
        fail("fail to decode want yaml file: ${e.path}: ${d.errorMessage}");
      }
      final v = d.value;
      if (v.type != Type.TYPE_OBJ) {
        fail("want yaml file is not a map: ${e.path}");
      }
      final wantError = v.obj["want_error"];
      if (wantError != null) {
        if (wantError.type != Type.TYPE_BOOL) {
          fail("want_error is not a bool: ${e.path}");
        }
        testcases[key]!.wantError = wantError.bool_2;
        return;
      }
      final wantValue = v.obj["want_value"];
      if (wantValue != null) {
        testcases[key]!.wantValue = wantValue;
        return;
      }
    }
  });

  final names = testcases.keys.toList()..sort();
  for (final name in names) {
    final testcase = testcases[name]!;
    test(name, () {
      final sut = Evaluator();
      final got = sut.evaluate(EvaluateInput(source: testcase.yamlInput));
      if (testcase.wantError ?? false) {
        expect(got.status, isNot(equals(EvaluateOutput_Status.OK)));
      } else {
        expect(got.status, (EvaluateOutput_Status.OK));
        final msg = _checkEqual([], testcase.wantValue!, got.value);
        if (msg != null) {
          fail(msg);
        }
      }
    });
  }
}

String? _checkEqual(List<String> path, Value want, Value got) {
  if (want.type != got.type) {
    return "type mismatch: /${path.join("/")}: ${want.type} != ${got.type}";
  }
  switch (want.type) {
    case Type.TYPE_NULL:
      return null;
    case Type.TYPE_BOOL:
      if (want.bool_2 != got.bool_2) {
        return "bool mismatch: /${path.join("/")}: ${want.bool_2} != ${got.bool_2}";
      }
      return null;
    case Type.TYPE_NUM:
      if (want.num != got.num) {
        return "num mismatch: /${path.join("/")}: ${want.num} != ${got.num}";
      }
      return null;
    case Type.TYPE_STR:
      if (want.str != got.str) {
        return "str mismatch: /${path.join("/")}: ${want.str} != ${got.str}";
      }
      return null;
    case Type.TYPE_ARR:
      if (want.arr.length != got.arr.length) {
        return "arr length mismatch: /${path.join("/")}: ${want.arr.length} != ${got.arr.length}";
      }
      for (var i = 0; i < want.arr.length; i++) {
        final msg =
            _checkEqual([...path, i.toString()], want.arr[i], got.arr[i]);
        if (msg != null) {
          return msg;
        }
      }
      return null;
    case Type.TYPE_OBJ:
      if (!ListEquality().equals(
          want.obj.keys.toList()..sort(), got.obj.keys.toList()..sort())) {
        return "obj key mismatch: /${path.join("/")}: ${want.obj.keys.toList()..sort()} != ${got.obj.keys.toList()..sort()}";
      }
      for (var key in want.obj.keys) {
        expect(got.obj.containsKey(key), isTrue);
        final msg = _checkEqual(
            [...path, key.toString()], want.obj[key]!, got.obj[key]!);
        if (msg != null) {
          return msg;
        }
      }
      return null;
    default:
      return "unknown type: /${path.join("/")}: ${want.type}";
  }
}

File: php/tests/Eval_/EvaluatorTest.php
<?php

namespace Jumpaku\Eyamluate\Tests\Eval_;

use Jumpaku\Eyamluate\Eval_\BaseEvaluator;
use Jumpaku\Eyamluate\Eval_\Evaluator;
use Jumpaku\Eyamluate\Eval_\PBEvaluateInput;
use Jumpaku\Eyamluate\Eval_\PBEvaluateOutput;
use Jumpaku\Eyamluate\Yaml\Decoder;
use Jumpaku\Eyamluate\Yaml\PBDecodeInput;
use Jumpaku\Eyamluate\Yaml\PBType;
use Jumpaku\Eyamluate\Yaml\PBValue;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Yaml\Yaml;


class EvaluatorTest extends TestCase
{
    /** @var array */
    private $testcases = [];

    protected function setUp(): void
    {
        $dataDir = join(DIRECTORY_SEPARATOR, [__DIR__, "..", "..", "..", "testdata"]);
        $filePaths = [$dataDir];
        while (count($filePaths) > 0) {
            $path = array_shift($filePaths);
            if ($path == "." || $path == "..") {
                continue;
            }
            if (is_dir($path)) {
                foreach (scandir($path) as $child) {
                    if ($child == "." || $child == "..") {
                        continue;
                    }
                    $filePaths[] = join(DIRECTORY_SEPARATOR, [$path, $child]);
                }
                continue;
            }
            if (str_ends_with($path, ".in.yaml")) {
                $key = mb_strimwidth($path, 0, strlen($path) - strlen(".in.yaml"));
                if (!array_key_exists($key, $this->testcases)) {
                    $this->testcases[$key] = new EvaluatorTestcase;
                }
                $this->testcases[$key]->inputYaml = file_get_contents($path);
            } elseif (str_ends_with($path, ".want.yaml")) {
                $key = mb_strimwidth($path, 0, strlen($path) - strlen(".want.yaml"));
                if (!array_key_exists($key, $this->testcases)) {
                    $this->testcases[$key] = new EvaluatorTestcase;
                }
                $yaml = (new Decoder)->decode((new PBDecodeInput)->setYaml(file_get_contents($path)));
                if ($yaml->getIsError()) {
                    $this->fail($yaml->getErrorMessage());
                }
                /** @var PBValue $wantValue */
                $wantValue = $yaml->getValue()->getObj()["want_value"];
                if ($wantValue != null) {
                    $this->testcases[$key]->wantValue = $wantValue;
                }
                /** @var PBValue $wantError */
                $wantError = $yaml->getValue()->getObj()["want_error"];
                if ($wantError != null) {
                    $this->testcases[$key]->wantError = $wantError->getBool();
                }
            }
        }
    }

    public function testEvaluate()
    {
        $names = array_keys($this->testcases);
        foreach ($names as $name) {
            print_r($name . "\n");
            /** @var EvaluatorTestcase $t */
            $t = $this->testcases[$name];
            $sut = new BaseEvaluator();
            $got = $sut->evaluate((new PBEvaluateInput)->setSource($t->inputYaml));

            if ($t->wantError) {
                $this->assertNotEquals(PBEvaluateOutput\PBStatus::OK, $got->getStatus());
            } else {
                $this->assertEquals(PBEvaluateOutput\PBStatus::OK, $got->getStatus());
                $msg = $this->checkValueEqual([], $t->wantValue, $got->getValue());
                if ($msg != null) {
                    $this->fail($msg);
                }
            }

        }
    }

    /**
     * @param string[] $path
     * @param PBValue $want
     * @param PBValue $got
     * @return string|null
     */
    private function checkValueEqual(array $path, PBValue $want, PBValue $got): string|null
    {
        if ($want->getType() != $got->getType()) {
            return sprintf("type mismatch: /%s", join("/", $path));
        }
        switch ($want->getType()) {
            case PBType::TYPE_NULL:
                return null;
            case PBType::TYPE_BOOL:
                if ($want->getBool() != $got->getBool()) {
                    return sprintf("boolean mismatch want %s, got %s: /%s", $want->getBool(), $got->getBool(), join("/", $path));
                }
                return null;
            case PBType::TYPE_NUM:
                if ($want->getNum() != $got->getNum()) {
                    return sprintf("number mismatch want %s, got %s: /%s", $want->getNum(), $got->getNum(), join("/", $path));
                }
                return null;
            case PBType::TYPE_STR:
                if ($want->getStr() != $got->getStr()) {
                    return sprintf("string mismatch want %s, got %s: /%s", $want->getStr(), $got->getStr(), join("/", $path));
                }
                return null;
            case PBType::TYPE_ARR:
                if ($want->getArr()->count() != $got->getArr()->count()) {
                    return sprintf("array length mismatch want %d, got %d: /%s", $want->getArr()->count(), $got->getArr()->count(), join("/", $path));
                }
                foreach ($want->getArr() as $index => $wantItem) {
                    $gotItem = $got->getArr()[$index];
                    $msg = $this->checkValueEqual([...$path, (string)$index], $wantItem, $gotItem);
                    if ($msg != null) {
                        return $msg;
                    }
                }
                return null;
            case PBType::TYPE_OBJ:
                $wk = array_keys(iterator_to_array($want->getObj()));
                $gk = array_keys(iterator_to_array($got->getObj()));
                sort($wk);
                sort($gk);
                if ($wk != $gk) {
                    return sprintf("object keys mismatch want [%s], got [%s]: /%s", join(",", $wk), join(",", $gk), join("/", $path));
                }
                foreach ($want->getObj() as $key => $wantItem) {
                    /** @var PBValue $gotItem */
                    $gotItem = $got->getObj()[$key];
                    $msg = $this->checkValueEqual([...$path, $key], $wantItem, $gotItem);
                    if ($msg != null) {
                        return $msg;
                    }
                }
                return null;
            default:
                assert(false, sprintf("unexpected type: %s", PBType::name($want->getType())));
        }
    }
}